src/download.c

```diff
@@ -74,252 +73,107 @@ static struct list *failed = NULL;
  */
 #define HASH_TO_KEY(hash) (HASH_VALUE(hash[0]) << 4 | HASH_VALUE(hash[1]))
 
-static struct hashmap *swupd_curl_hashmap = NULL;
+struct swupd_curl_parallel_handle {
+	int timeout; /* Retry timeout */
+	size_t mcurl_size, max_xfer, max_xfer_bottom; /* hysteresis parameters */
 
-/* try to insert the file into the hashmap download queue
- * returns 1 if no download is needed
- * returns 0 if download is needed
- * returns -1 if error */
-static int swupd_curl_hashmap_insert(struct file *file)
-{
-	char *tar_dotfile;
-	char *targetfile;
-	struct stat stat;
+	CURLM *mcurl; /* Curl handle */
+	struct list *failed; /* List of failed downloads */
+	struct hashmap *curl_hashmap; /* Hashmap mentioned above */
+	swupd_curl_download_cb success_cb; /* Callback to sucess function */
+};
 
-	if (!hashmap_put(swupd_curl_hashmap, file)) {
-		// hash already in download queue
-		return 1;
-	}
-
-	// if valid target file is already here, no need to download
-	string_or_die(&targetfile, "%s/staged/%s", state_dir, file->hash);
-
-	if (lstat(targetfile, &stat) == 0) {
-		/* file exists */
-		if (verify_file(file, targetfile)) {
-			/* hash matches, no download necessary */
-			free_string(&targetfile);
-			return 1;
-		} else {
-			/* hash mismatch, remove the staged file to enable re-download */
-			unlink(targetfile);
-		}
-	}
-
-	free_string(&targetfile);
-
-	// hash not in queue and not present in staged
+/*
+ * Struct represinting a single file being downloaded.
+ */
+struct multi_curl_file {
+	struct curl_file file; /* Curl file information */
 
-	// clean up in case any prior download failed in a partial state
-	string_or_die(&tar_dotfile, "%s/download/.%s.tar", state_dir, file->hash);
-	unlink(tar_dotfile);
-	free_string(&tar_dotfile);
+	char retries; /* Number of retried performed so far */
+	CURL *curl; /* curl handle if downloading */
+	char *url; /* The url to be downloaded from */
+	size_t hash_key; /* hash_key of this file */
+	const char *hash; /* Unique identifier of this file. */
 
-	return 0;
-}
-
-/* hysteresis thresholds */
-static size_t MAX_XFER = 25;
-static size_t MAX_XFER_BOTTOM = 15;
+	void *data; /* user's data */
+};
 
 static bool file_hash_cmp(const void *a, const void *b)
 {
-	const struct file *fa = a;
-	const struct file *fb = b;
+	const struct multi_curl_file *fa = a;
+	const struct multi_curl_file *fb = b;
 
-	return hash_equal(fa->hash, fb->hash);
+	return strcmp(fa->hash, fb->hash) == 0;
```

Comment:
Reason for this change?

Reply:
hash is supposed to be an abstract data type, so I am very much with @matthewrsj here, and would like this reverted.

Reply:
Yes when I did it, but not right now. I'll fix that